### PR TITLE
[docs] Distributed inference

### LINKED ISSUE #3010: [Docs] Distributed inference with multiple GPUs
Originally posted by @patrickvonplaten in https://github.com/huggingface/diffusers/issues/2977#issuecomment-1498832883

### PR DESCRIPTION
**Pull request by stevhliu**
Fixes #3010 by adding distributed inference to the docs with the option to use either `torch.distributed` or 🤗 Accelerate.

### CODE REVIEW

File: docs/source/en/training/distributed_inference.mdx [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# Distributed inference with multiple GPUs
+
+On distributed setups, you can run inference across multiple GPUs with [PyTorch Distributed](https://pytorch.org/tutorials/beginner/dist_overview.html) or 🤗 [Accelerate](https://huggingface.co/docs/accelerate/index), which is useful for generating multiple prompts in parallel.
```
Comment (sayakpaul) [inline docs/source/en/training/distributed_inference.mdx:3]:
```suggestion
On distributed setups, you can run inference across multiple GPUs with [PyTorch Distributed](https://pytorch.org/tutorials/beginner/dist_overview.html) or 🤗 [Accelerate](https://huggingface.co/docs/accelerate/index), which is useful for generating with multiple prompts in parallel.
```

File: docs/source/en/training/distributed_inference.mdx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+sd = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
+```
+
+You'll want to create a function to run inference; [`init_process_group`](https://pytorch.org/docs/stable/distributed.html?highlight=init_process_group#torch.distributed.init_process_group) handles creating a distributed environment with the type of backend to use, the `rank` of the current process, and the `world_size` or the number of processes participating. If you're running inference in parallel over 2 GPUs, then the `world_size` would be 2.
+
+Move the [`DiffusionPipeline`] to `rank` and use `get_rank` to assign a GPU to each process, where each process handles a different prompt:
+
+```py
+def run_inference(rank, world_size):
+    dist.init_process_group("gloo", rank=rank, world_size=world_size)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply docs/source/en/training/distributed_inference.mdx:30]: In general, for all things GPU, I have found `nccl` to be a much better backend.

File: docs/source/en/training/distributed_inference.mdx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+This guide will show you how to use PyTorch Distributed and 🤗 Accelerate for distributed inference.
+
+## PyTorch Distributed
+
+PyTorch supports [`DistributedDataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html) which enables faster data parallelism.
+
+To start, create a Python file and import `torch.distributed` and `torch.multiprocessing` to set up the distributed process group and to spawn the processes for inference on each GPU. You should also initialize a [`DiffusionPipeline`]:
+
+```py
+#!/usr/bin/env python3
```
Comment (sayakpaul) [inline docs/source/en/training/distributed_inference.mdx:14]: Do we need this?

**Review by sayakpaul — APPROVED** [review]
Looks amazing! 

I think we need to add licensing in the inference file. And then we should be good.

**Comment by sayakpaul** [comment]
@muellerzr do you mind taking a look here as well?

File: docs/source/en/training/distributed_inference.mdx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+
+if __name__ == "__main__":
+    main()
+```
+
+Once you've completed the inference script, run it like:
+
+```bash
+torchrun run_distributed.py
```
Comment (muellerzr) [inline docs/source/en/training/distributed_inference.mdx:58]: A small nit, it would be good to include here (and in the accelerate cli example) how to specify the two workers mentioned, e.g. the `--nproc_per_node` param and in `accelerate` the `accelerate launch --num_processes 2`

File: docs/source/en/training/distributed_inference.mdx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    sd.to(state.device)
+
+    if state.process_index == 0:
+        prompt = "a dog"
+    elif state.process_index == 1:
+        prompt = "a cat"
+
+    image = sd(prompt).images[0]
+    image.save(f"./{'_'.join(prompt)}.png")
```
Comment (muellerzr) [inline docs/source/en/training/distributed_inference.mdx:86]: I feel like we may worry about deadlocking the CPU here by accident no? (Haven't fully tried this yet, just thinking of how we do things in Accelerate that's *close to this*). So e.g. might be better to have a `main_process_first` during the `save` portion to be safe. (Again, something I'd need to play with, just voicing my thoughts here)

**Review by muellerzr — APPROVED** [review]
This is *amazing*, and definitely would like to steal some (or much) of this over in the Accelerate docs as well, as distributed inference is something we've been asked for guides on I haven't gotten to yet :) 

Great work!

**Review by pcuenca — APPROVED** [review]
Looks great to me! In my opinion this could be merged once @muellerzr's comments are clarified :)

**Review by patrickvonplaten — APPROVED** [review]
Cool!

**Comment by stevhliu** [comment]
@muellerzr just made distributed inference even easier with https://github.com/huggingface/accelerate/pull/1410 😮‍💨

Once it gets merged, I'll update the code example here with the `split_between_processes` utility!

**Comment by muellerzr** [comment]
We've now merged it in main, and borrowed some from the excellent example here 😉 😄

File: docs/source/en/training/distributed_inference.mdx [hunk 6]
```diff
@@ -0,0 +1,91 @@
+# Distributed inference with multiple GPUs
+
+On distributed setups, you can run inference across multiple GPUs with [PyTorch Distributed](https://pytorch.org/tutorials/beginner/dist_overview.html) or 🤗 [Accelerate](https://huggingface.co/docs/accelerate/index), which is useful for generating with multiple prompts in parallel.
+
+This guide will show you how to use PyTorch Distributed and 🤗 Accelerate for distributed inference.
+
+## PyTorch Distributed
+
+PyTorch supports [`DistributedDataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html) which enables faster data parallelism.
```
Comment (pcuenca) [inline docs/source/en/training/distributed_inference.mdx:9]:
```suggestion
PyTorch supports [`DistributedDataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html) which enables data parallelism.
```

Why "faster"?

File: docs/source/en/training/distributed_inference.mdx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    main()
+```
+
+Once you've completed the inference script, use the `--nproc_per_node` argument to specify the number of GPUs to use and call `torchrun` to run the script:
+
+```bash
+torchrun run_distributed.py --nproc_per_node=2
+```
+
+## 🤗 Accelerate
```
Comment (pcuenca) [inline docs/source/en/training/distributed_inference.mdx:60]: Why not put this first rather than DDP? It's so much easier! :)

File: docs/source/en/training/distributed_inference.mdx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+🤗 [Accelerate](https://huggingface.co/docs/accelerate/index) is a library designed to make it easy to train or run inference across distributed setups. It simplifies the process of setting up the distributed environment, allowing you to focus on your PyTorch code.
+
+To begin, create a Python file and initialize an [`accelerate.PartialState`] to create a distributed environment; your setup is automatically detected so you don't need to explicitly define the `rank` or `world_size`. Move the [`DiffusionPipeline`] to `distributed_state.device` to assign a GPU to each process.
+
+<Tip>
+
+To learn more about how to perform distributed inference, for example, if you have an odd number of GPUs or want to pad the prompts to the same length, take a look at the [Distributed Inference with 🤗 Accelerate](https://huggingface.co/docs/accelerate/en/usage_guides/distributed_inference#distributed-inference-with-accelerate) guide.
```
Comment (pcuenca) [inline docs/source/en/training/distributed_inference.mdx:68]: I'd maybe put this last as a comment without mentioning the odd number of GPUs or padding, I found it confusing. After checking the docs, turns out it has to do with sending the same number of prompts to each GPU, not really about them having the same length. I think that `accelerate` does the right thing by default, so I'd focus on that path.

File: docs/source/en/training/distributed_inference.mdx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+```py
+from accelerate import PartialState
+from diffusers import DiffusionPipeline
+
+pipeline = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
+distributed_state = PartialState()
+pipeline.to(distributed_state.device)
+
+with distributed_state.split_across_processes(["a dog", "a cat"]) as prompt:
```
Comment (pcuenca) [inline docs/source/en/training/distributed_inference.mdx:82]:
```suggestion
with distributed_state.split_between_processes(["a dog", "a cat"]) as prompt:
```

;)

File: docs/source/en/training/distributed_inference.mdx [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    result = pipe(prompt).images[0]
```
Comment (pcuenca) [inline docs/source/en/training/distributed_inference.mdx:83]:
```suggestion
    result = pipeline(prompt).images[0]
```

File: docs/source/en/training/distributed_inference.mdx [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    result.save(f"result_{distributed_state.rank}.png")
+```
+
+Use the `--num_processes` argument to specify the number of GPUs to use, and call `accelerate launch` to run the script:
+
```
Comment (pcuenca) [inline docs/source/en/training/distributed_inference.mdx:88]: Are there any particular considerations we need to take into account when using `accelerate config`?

**Review by pcuenca — COMMENTED** [review]
Great work!

File: docs/source/en/training/distributed_inference.mdx [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+from accelerate import PartialState
+from diffusers import DiffusionPipeline
+
+pipeline = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
+distributed_state = PartialState()
+pipeline.to(distributed_state.device)
+
+with distributed_state.split_across_processes(["a dog", "a cat"]) as prompt:
+    result = pipe(prompt).images[0]
+    result.save(f"result_{distributed_state.rank}.png")
```
Comment (pcuenca) [inline docs/source/en/training/distributed_inference.mdx:84]:
```suggestion
    result.save(f"result_{distributed_state.process_index}.png")
```

I couldn't find the `rank` property.

File: docs/source/en/training/distributed_inference.mdx (hunk 11 above)
Comment (muellerzr) [reply docs/source/en/training/distributed_inference.mdx:88]: Nope, just that you could also _not_ use config if desired

File: docs/source/en/training/distributed_inference.mdx (hunk 6 above)
Comment (stevhliu) [reply docs/source/en/training/distributed_inference.mdx:9]: I think I was trying to describe how it's different from just `DataParallel`, but I guess it doesn't really make sense if I don't mention it at all! 😅 Better to just remove it and keep it simple :)

**Review by patrickvonplaten — APPROVED** [review]
Very cool!
